File: archive/sdk/python/paw/modules/bank.py
"""Bank module for token operations."""

from typing import List, Optional
from ..types import Coin, SendParams, TxResult, GasOptions


class BankModule:
    """Bank module for managing tokens."""

    def __init__(self, client):
        """Initialize bank module."""
        self.client = client

    async def get_balance(self, address: str, denom: str) -> Optional[Coin]:
        """Get balance for a specific denom."""
        try:
            data = await self.client.get(f"/cosmos/bank/v1beta1/balances/{address}/{denom}")
            balance = data.get("balance")
            if balance:
                return Coin(denom=balance["denom"], amount=balance["amount"])
            return None
        except Exception:
            return None

    async def get_all_balances(self, address: str) -> List[Coin]:
        """Get all balances for an address."""
        try:
            data = await self.client.get(f"/cosmos/bank/v1beta1/balances/{address}")
            balances = data.get("balances", [])
            return [Coin(denom=b["denom"], amount=b["amount"]) for b in balances]
        except Exception:
            return []

    async def send(
        self,
        sender: str,
        params: SendParams,
        options: Optional[GasOptions] = None
    ) -> TxResult:
        """Send tokens to another address."""
        message = {
            "@type": "/cosmos.bank.v1beta1.MsgSend",
            "from_address": sender,
            "to_address": params.recipient,
            "amount": [{"denom": params.denom, "amount": params.amount}]
        }

        if options is None:
            options = GasOptions(memo=params.memo)

        return await self.client.tx_builder.sign_and_broadcast([message], options)

    def format_balance(self, balance: Coin, decimals: int = 6) -> str:
        """Format balance for display."""
        amount = int(balance.amount) / (10 ** decimals)
        denom_upper = balance.denom.replace("u", "").upper()
        return f"{amount:.{decimals}f} {denom_upper}"

File: archive/sdk/javascript/src/modules/bank.ts
import { PawClient } from '../client';
import { SendParams, TxResult, GasOptions } from '../types';
import { Coin } from '@cosmjs/stargate';

export class BankModule {
  constructor(private client: PawClient) {}

  /**
   * Get account balance for a specific denom
   */
  async getBalance(address: string, denom: string): Promise<Coin | null> {
    const client = this.client.getClient();
    return await client.getBalance(address, denom);
  }

  /**
   * Get all account balances
   */
  async getAllBalances(address: string): Promise<readonly Coin[]> {
    const client = this.client.getClient();
    return await client.getAllBalances(address);
  }

  /**
   * Send tokens to another address
   */
  async send(
    senderAddress: string,
    params: SendParams,
    options?: GasOptions
  ): Promise<TxResult> {
    const denom = params.denom || 'upaw';
    const message = {
      typeUrl: '/cosmos.bank.v1beta1.MsgSend',
      value: {
        fromAddress: senderAddress,
        toAddress: params.recipient,
        amount: [{ denom, amount: params.amount }]
      }
    };

    const txBuilder = this.client.getTxBuilder();
    return await txBuilder.signAndBroadcast(senderAddress, [message], {
      ...options,
      memo: params.memo
    });
  }

  /**
   * Multi-send tokens to multiple recipients
   */
  async multiSend(
    senderAddress: string,
    recipients: Array<{ address: string; amount: string; denom?: string }>,
    options?: GasOptions
  ): Promise<TxResult> {
    const outputs = recipients.map(recipient => ({
      address: recipient.address,
      coins: [{ denom: recipient.denom || 'upaw', amount: recipient.amount }]
    }));

    const message = {
      typeUrl: '/cosmos.bank.v1beta1.MsgMultiSend',
      value: {
        inputs: [{
          address: senderAddress,
          coins: outputs.flatMap(o => o.coins)
        }],
        outputs
      }
    };

    const txBuilder = this.client.getTxBuilder();
    return await txBuilder.signAndBroadcast(senderAddress, [message], options);
  }

  /**
   * Get total supply of a denom
   */
  async getTotalSupply(denom: string): Promise<Coin | null> {
    try {
      const config = this.client.getConfig();
      const restEndpoint = config.restEndpoint || config.rpcEndpoint.replace(':26657', ':1317');

      const response = await fetch(`${restEndpoint}/cosmos/bank/v1beta1/supply/by_denom?denom=${denom}`);
      if (!response.ok) {
        return null;
      }

      const data = await response.json() as { amount?: Coin };
      return data.amount || null;
    } catch (error) {
      console.error('Error fetching total supply:', error);
      return null;
    }
  }

  /**
   * Get all denoms
   */
  async getAllDenoms(): Promise<string[]> {
    try {
      const config = this.client.getConfig();
      const restEndpoint = config.restEndpoint || config.rpcEndpoint.replace(':26657', ':1317');

      const response = await fetch(`${restEndpoint}/cosmos/bank/v1beta1/supply`);
      if (!response.ok) {
        return [];
      }

      const data = await response.json() as { supply?: Coin[] };
      return (data.supply || []).map((coin: Coin) => coin.denom);
    } catch (error) {
      console.error('Error fetching all denoms:', error);
      return [];
    }
  }

  /**
   * Get spendable balance (available for spending)
   */
  async getSpendableBalance(address: string, denom: string): Promise<Coin | null> {
    // For now, spendable balance is the same as regular balance
    // In the future, this might account for locked/vesting tokens
    return await this.getBalance(address, denom);
  }

  /**
   * Format balance for display
   */
  formatBalance(balance: Coin, decimals: number = 6): string {
    const amount = parseInt(balance.amount) / Math.pow(10, decimals);
    return `${amount.toFixed(decimals)} ${balance.denom.replace('u', '').toUpperCase()}`;
  }
}
